### PR TITLE
Fix #288 nullable option not working for reference

### LINKED ISSUE #288: Reference/Embbeded column with nullable=true couldn't be set to null by setMethod
I think for sth like

```php
/**
 * 
 * @MongoDB\ReferenceOne(targetDocument="Sth", nullable=true)
 */
protected $test;
```

should produce

```php
public function setTest(Sth $test = null)
    {
        $this->test = $test;
    }
```

instead of 

```php
public function setTest(Sth $test)
    {
        $this->test = $test;
    }
```

According to documentation (http://php.net/manual/en/language.oop5.typehinting.php) this is the only way for accepting "null" as parameter when type hinting. Of course if there is "nullable=false" generator could create method as now.

### PR DESCRIPTION
**Pull request by ebuildy**
Reference/Embbeded column with nullable=true couldn't be set to null by setMethod

### CODE REVIEW

File: lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php [hunk 1]
```diff
@@ -461,7 +461,7 @@ private function hasMethod($method, ClassMetadataInfo $metadata)
     {
         if ($this->extendsClass()) {
             // don't generate method if its already on the base class.
-            $reflClass = new \ReflectionClass($this->getClassToExtend());
+            $reflClass = new \ReflectionClass($this->getClassTogenerateDocumentStubMethodExtend());
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php]: Does this method exist?

File: lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php [hunk 2]
```diff
@@ -667,6 +667,7 @@ private function generateDocumentStubMethods(ClassMetadataInfo $metadata)
                     $methods[] = $code;
                 }
             } elseif ($fieldMapping['type'] === ClassMetadataInfo::ONE) {
+                $nullable = $this->isAssociationIsNullable($fieldMapping) ? 'null' : null;
                 if ($code = $this->generateDocumentStubMethod($metadata, 'set', $fieldMapping['fieldName'], isset($fieldMapping['targetDocument']) ? $fieldMapping['targetDocument'] : null)) {
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php]: It seems like you haven't passed ``$nullable`` here

File: lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php [hunk 3]
```diff
@@ -688,6 +689,16 @@ private function generateDocumentStubMethods(ClassMetadataInfo $metadata)
 
         return implode("\n\n", $methods);
     }
+    
+    /**
+     * @param array $fieldMapping
+     *
+     * @return bool
+     */
+    protected function isAssociationIsNullable($fieldMapping)
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php]: there are 2 "is" in method name, also I believe this should be private

File: lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php [hunk 4]
```diff
@@ -771,12 +782,13 @@ private function generateDocumentStubMethod(ClassMetadataInfo $metadata, $type,
...
-            '<variableName>'   => $variableName,
-            '<methodName>'     => $methodName,
-            '<fieldName>'      => $fieldName,
+            '<description>'         => $description,
+            '<methodTypeHint>'      => $methodTypeHint,
+            '<variableType>'        => $variableType,
+            '<variableName>'        => $variableName,
+            '<methodName>'          => $methodName,
+            '<fieldName>'           => $fieldName,
+            '<variableDefault>'     => ($defaultValue !== null ) ? (' = '.$defaultValue) : '',
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php]: Can you please change concat to use spaces so ``' = ' . $defaultValue``?

File: lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php (hunk 3 above)
Comment (ebuildy) [inline lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php]: True, I copied the declaration from https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/Tools/EntityGenerator.php#L1225, but no problem to remove the extra is and make it private!

**Comment by ebuildy** [comment]
Thanks you for your feedbacks, I am very glad to help in a such important project!

File: lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php [hunk 5]
```diff
@@ -667,7 +667,8 @@ private function generateDocumentStubMethods(ClassMetadataInfo $metadata)
                     $methods[] = $code;
                 }
             } elseif ($fieldMapping['type'] === ClassMetadataInfo::ONE) {
-                if ($code = $this->generateDocumentStubMethod($metadata, 'set', $fieldMapping['fieldName'], isset($fieldMapping['targetDocument']) ? $fieldMapping['targetDocument'] : null)) {
+                $nullable = $this->isAssociationNullable($fieldMapping) ? 'null' : null;
+                if ($code = $this->generateDocumentStubMethod($metadata, 'set', $fieldMapping['fieldName'], isset($fieldMapping['targetDocument']) ? $fieldMapping['targetDocument'] : null), $nullable) {
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php]: ``$nullable`` is still in wrong place, take a look at failing tests in Travis. You can also run tests locally which eases life a lot :)

**Comment by malarzm** [comment]
@ebuildy thanks for the fixed, there's still one thing left as Travis is failing. Could you also squash your commits later?

File: lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php (hunk 5 above)
Comment (ebuildy) [inline lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php]: I am really sorry, I am not at home nowadays, and work from my wife computer ;-( without Php of course, am fixing it

Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php]: Thanks! Also props for strong will for working without PHP ;)

Comment (ebuildy) [inline lib/Doctrine/ODM/MongoDB/Tools/DocumentGenerator.php]: Github is a good place for that! And with good unit tests of course ;-)

Sorry for the loooonngg PR ^^ I will be more attentive next time.

**Comment by malarzm** [comment]
Merged manually in https://github.com/doctrine/mongodb-odm/commit/0de70ab260a7d0e5b872bb65e79e8ebc1c033b85 - thanks @ebuildy!

**Pull request closed without merging**